### PR TITLE
Ajax: Warn against automatic JSON-to-JSONP promotion

### PR DESCRIPTION
**Pull request by mgol**
Fixes gh-372
Ref jquery/jquery#1799
Ref jquery/jquery#3376
Ref jquery/jquery#4754

### CODE REVIEW

File: test/ajax.js [hunk 1]
```diff
@@ -19,12 +19,104 @@ QUnit.test( "jQuery.ajax() deprecations on jqXHR", function( assert ) {
...
+
+		assert.expect( 5 );
+
+		var done = assert.async(),
+			tests = [
+				function() {
+					return expectNoWarning( assert, "dataType: \"json\"",
+						function() {
+							return jQuery.ajax( {
+								url: url( "data/null.json" ),
```
Comment (mgol) [inline test/ajax.js:45]: This file just contains `null` which is both valid JSON & JS. Quite handy!

File: test/ajax.js [hunk 2]
```diff
@@ -19,12 +19,104 @@ QUnit.test( "jQuery.ajax() deprecations on jqXHR", function( assert ) {
...
+				}
 			} );
-	} );
+		}
 
+		run( tests )
+			.then( function() {
+				done();
+			} );
+	} );
```
Comment (mgol) [inline test/ajax.js:119]: This is quite a lot of boilerplate but we don't have that many tests so I'm not sure if this is the moment to generalize it... If there's a way to simplify, I'm all ears. :)

File: test/index.html [hunk 3]
```diff
@@ -14,7 +14,7 @@
 	<!-- Load a jQuery and jquery-migrate plugin file based on URL -->
 	<script src="testinit.js"></script>
 	<script>
-		TestManager.loadProject( "jquery", "git" );
+		TestManager.loadProject( "jquery", "3.x-git" );
```
Comment (mgol) [inline test/index.html:17]: I can separate these changes if you want.

File: src/jquery/ajax.js [hunk 4]
```diff
@@ -21,4 +23,26 @@ jQuery.ajax = function( ) {
...
 
+// Only trigger the logic in jQuery <4 as the JSON-to-JSONP auto-promotion
+// behavior is gone in jQuery 4.0 and as it has security implications, we don't
+// want to restore the legacy behavior.
+if ( !jQueryVersionSince( "4.0.0" ) ) {
+
+	// Register this prefilter before the jQuery one. Otherwise, a promoted
+	// request is transformed into one with the script dataType and we can't
+	// catch it anymore.
+	jQuery.ajaxPrefilter( "+json", function( s ) {
```
Comment (mgol) [inline src/jquery/ajax.js:34]: The special `+` handling is not documented at https://api.jquery.com/jQuery.ajaxPrefilter/, I found it by reading source. 🙈

File: test/ajax.js (hunk 2 above)
Comment (dmethvin) [reply test/ajax.js:119]: I'm fine with it, if it really gets out of hand we can refactor some other time.

File: test/migrate.js [hunk 5]
```diff
@@ -1,30 +1,45 @@
...
+		// Failure; use deepEqual to show the warnings that *were* generated and the expectation
+		} else {
+			assert.deepEqual( jQuery.migrateWarnings,
+				"<warnings: " + ( expected || "1+" ) + ">", name + ": warned"
+			);
+		}
+	}
+
+	if ( result && result.then ) {
+		return jQuery.when(
```
Comment (mgol) [inline test/migrate.js:35]: Note to self - `jQuery.when` is unavailable in jQuery 4.x Slim. Perhaps it's better to use a Promise polyfill like [native-promise-only](https://www.npmjs.com/package/native-promise-only) that we already use in Core tests.

**Comment by mgol** [comment]
@dmethvin I added one commit switching `jQuery.when` to `Promise.resolve` using a `Promise` polyfill. This works with the Core 4.x slim build. Let me know if this is still good to land.

**Pull request merged**